sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,375 @@
+    """
+    def __init__(self, n_features_to_select=0.5, use_redundancy=True,
+                 categorical_features=False, categorical_target=False,
+                 n_neighbors=3, copy=True):
+        self.n_features_to_select = n_features_to_select
+        self.use_redundancy = use_redundancy
+        self.categorical_features = categorical_features
+        self.categorical_target = categorical_target
+        self.n_neighbors = n_neighbors
+        self.copy = copy
+
+    def _get_support_mask(self):
+        n_features_to_select = self._get_n_features_to_select()
+        support = np.zeros_like(self.ranks_, dtype=bool)
+        support[self.ranks_ < n_features_to_select] = True
+        return support
+
+    def _get_n_features_to_select(self):
+        if self.n_features_to_select >= 1:
+            return int(self.n_features_to_select)
+        elif 0 < self.n_features_to_select < 1:
+            return max(1, int(self.n_features_to_select * self._n_features))
+        else:
+            raise ValueError("`n_features_to_select` must be positive.")
+
+    def fit(self, X, y):
+        """Find the best features according to mutual information criterion.
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self
+        """
+        X, y = check_X_y(X, y, accept_sparse='csc',
+                         y_numeric=not self.categorical_target)
+
+        n_features = X.shape[1]
+        self._n_features = n_features
+        n_features_to_select = self._get_n_features_to_select()
+
+        if isinstance(self.categorical_features, bool):
+            categorical_features = np.empty(n_features, dtype=bool)
+            categorical_features.fill(self.categorical_features)
+        else:
+            categorical_features = np.asarray(self.categorical_features)
+
+        if (not np.all(categorical_features) and
+                (self.use_redundancy or not self.categorical_target)):
+            X = scale(X, with_mean=False, copy=self.copy)
+
+        if not self.categorical_target:
+            y_std = np.std(y)
+            if y_std > 0:
+                y = y / y_std
+
+        if issparse(X):
+            iterate_columns = _iterate_csc_columns
+        else:
+            iterate_columns = _iterate_dense_columns
+
+        relevance = np.empty(n_features)
+        for i, xi in enumerate(iterate_columns(X)):
+            relevance[i] = _compute_mi(xi, y, categorical_features[i],
+                                       self.categorical_target,
+                                       self.n_neighbors)
+
+        if not self.use_redundancy:
+            self.ranks_ = np.argsort(-relevance)
+            return self
+
+        ranks = np.empty(n_features, dtype=int)
+        ranks.fill(n_features - 1)
```

Comment:
Yes I understood that. It was merely a stylistic comment.

    n_features = 5
    ranks = np.empty(5, dtype=int)
    ranks.fill(4)
    print(ranks)
    ranks = np.ones(5, dtype=int) * (4)